cardinal/server/tx.go

```diff
@@ -1,127 +1,144 @@
-				return middleware.Error(http.StatusUnauthorized, eris.ToString(err, true)), nil
+		var signerAddress string
+		if msgType.Name() == ecs.CreatePersonaMsg.Name() {
+			// don't need to check the cast bc we already validated this above
+			createPersonaMsg, _ := msg.(ecs.CreatePersona)
+			signerAddress = createPersonaMsg.SignerAddress
+		} else {
+			signerAddress, err = s.eng.GetSignerForPersonaTag(tx.PersonaTag, 0)
+			if err != nil {
+				return fiber.NewError(fiber.StatusBadRequest, "could not get signer for persona: "+err.Error())
 			}
-			return middleware.Error(http.StatusInternalServerError, eris.ToJSON(err, true)), nil
 		}
-
-		txReply, err := handler.generateCreatePersonaResponseFromPayload(payload, sp, ecs.CreatePersonaMsg)
-		if err != nil {
-			return nil, err
+		if !s.disableSignatureVerification {
+			err = validateTransaction(tx, signerAddress, s.eng.Namespace().String(), tx.IsSystemTransaction())
+			if err != nil {
+				return fiber.NewError(fiber.StatusBadRequest, "failed to validate transaction: "+err.Error())
+			}
+			if err = s.eng.UseNonce(signerAddress, tx.Nonce); err != nil {
+				return fiber.NewError(fiber.StatusInternalServerError, "failed to use nonce: "+err.Error())
+			}
 		}
-		return &txReply, nil
-	})
 
-	api.RegisterOperation("POST", "/tx/game/{txType}", gameHandler)
-	api.RegisterOperation("POST", "/tx/persona/create-persona", createPersonaHandler)
+		tick, hash := s.eng.AddTransaction(msgType.ID(), msg, tx)
 
-	return nil
+		return ctx.JSON(&TransactionReply{
+			TxHash: string(hash),
+			Tick:   tick,
+		})
+	}
 }
 
-// submitTransaction submits a transaction to the game world, as well as the blockchain.
-func (handler *Handler) submitTransaction(txVal any, tx message.Message, sp *sign.Transaction,
-) (*TransactionReply, error) {
-	log.Debug().Msgf("submitting transaction %d: %v", tx.ID(), txVal)
-	tick, txHash := handler.w.AddTransaction(tx.ID(), txVal, sp)
-	txReply := &TransactionReply{
-		TxHash: string(txHash),
-		Tick:   tick,
+func getMessageAndTx(body []byte, mt message.Message) (*sign.Transaction, any, error) {
+	if len(body) == 0 {
+		return nil, nil, fiber.NewError(fiber.StatusBadRequest, "request body was empty")
+	}
+	tx, err := decodeTransaction(body)
+	if err != nil {
+		return nil, nil, fiber.NewError(fiber.StatusBadRequest, "transaction data malformed: "+err.Error())
 	}
-	return txReply, nil
+	msg, err := mt.Decode(tx.Body)
+	if err != nil {
+		return nil, nil, fiber.NewError(
+			fiber.StatusBadRequest,
+			"failed to decode message from transaction body: "+err.Error(),
+		)
+	}
+
+	return tx, msg, nil
+}
+
+func validateTransaction(tx *sign.Transaction, signerAddr, namespace string, systemTx bool) error {
+	if tx.PersonaTag == "" {
+		return ErrNoPersonaTag
+	}
+	if tx.Namespace != namespace {
+		return fmt.Errorf("expected %q got %q: %w", namespace, tx.Namespace, ErrWrongNamespace)
```

Comment:
these need to be wrapped in eris.

Reply:
[wrap validateTransaction errors with eris](https://github.com/Argus-Labs/world-engine/pull/544/commits/e93ee1d2e6c2ce561c3f4afbca51eeedd5ac33bd) fixed